src/librustc_trans/mir/mod.rs

```diff
@@ -102,12 +101,67 @@ pub struct MirContext<'bcx, 'tcx:'bcx> {
-    pub fn debug_loc(&self, source_info: mir::SourceInfo) -> DebugLoc {
-        DebugLoc::ScopeAt(self.scopes[source_info.scope], source_info.span)
+    pub fn debug_loc(&mut self, source_info: mir::SourceInfo) -> DebugLoc {
+        // Bail out if debug info emission is not enabled.
+        match self.fcx.debug_context {
+            FunctionDebugContext::DebugInfoDisabled |
+            FunctionDebugContext::FunctionWithoutDebugInfo => {
+                // Can't return DebugLoc::None here because intrinsic::trans_intrinsic_call()
+                // relies on debug location to obtain span of the call site.
+                return DebugLoc::ScopeAt(self.scopes[source_info.scope].scope_metadata,
+                                         source_info.span);
+            }
+            FunctionDebugContext::RegularContext(_) =>{}
+        }
+
+        // In order to have good line stepping behavior in a debugger, we overwrite debug
+        // locations of macro expansions with that of the outermost expansion site
+        // (unless the crate is being compiled with `-Z debug-macros`).
+        if source_info.span.expn_id == NO_EXPANSION ||
+           self.fcx.ccx.sess().opts.debugging_opts.debug_macros {
+            let scope_metadata = self.scope_metadata_for_span(source_info.scope, source_info.span);
+            DebugLoc::ScopeAt(scope_metadata, source_info.span)
+        } else {
+            // Walk up the macro expansion chain till we reach a non-expanded span.
+            let mut scope_id = source_info.scope;
+            let mut span = source_info.span;
+            let cm = self.fcx.ccx.sess().codemap();
+            while span.expn_id != NO_EXPANSION {
+                span = cm.source_callsite(span);
+            }
+            // Ditto for scope
+            while self.mir.visibility_scopes[scope_id].span.expn_id != NO_EXPANSION {
+                let parent_scope_id = self.mir.visibility_scopes[scope_id].parent_scope;
+                scope_id = match parent_scope_id {
+                    Some(id) => id,
+                    None => break // Can happen if the current function is defined inside a macro.
+                }
+            }
+            let scope_metadata = self.scope_metadata_for_span(scope_id, span);
+            DebugLoc::ScopeAt(scope_metadata, span)
+        }
+    }
+
+    // DILocations inherit source file name from the parent DIScope.  Due to macro expansions
+    // it may so happen that the current span belongs to a different file than the DIScope
+    // corresponding to span's containing visibility scope.  If so, we need to create a DIScope
+    // "extension" into that file.
+    fn scope_metadata_for_span(&self, scope_id: mir::VisibilityScope, span: Span)
+                               -> llvm::debuginfo::DIScope {
+        let scope_metadata = self.scopes[scope_id].scope_metadata;
+        if span.lo < self.scopes[scope_id].start_pos ||
+           span.lo > self.scopes[scope_id].end_pos {
```

Comment:
Debug locations are points, not ranges, and we are using span.lo, so that's what I am checking.
Besides, there's nothing to prevent a span from straddling a file boundary (e.g. a buggy syntax extension could generate one).